### PR TITLE
Apply custom scalar's parse fn to resolved variables 

### PR DESCRIPTION
**Pull request by annapawlicka**
This PR fixes https://github.com/walmartlabs/lacinia/issues/34

Variables can be custom scalars, and so we should treat them as such when their value is resolved.

### CODE REVIEW

File: test/com/walmartlabs/lacinia/custom_scalars_test.clj [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+(deftest custom-scalar-query
+  (let [q "{ now { date }}"]
+    (is (= {:data {:now {:date "A long time ago"}}}
+           (execute default-schema q nil nil)))))
+
+(deftest custom-scalars
+  (testing "custom scalars defined as conformers"
+    (let [parse-conformer (s/conformer
+                           (fn [x]
+                             (prn "qoooooooo")
```
Comment (bcarrell) [inline test/com/walmartlabs/lacinia/custom_scalars_test.clj]: Left a `prn` in there.

Comment (kgann) [inline test/com/walmartlabs/lacinia/custom_scalars_test.clj]: 👀

Comment (annapawlicka) [inline test/com/walmartlabs/lacinia/custom_scalars_test.clj]: Ooops haha

File: src/com/walmartlabs/lacinia/parser.clj [hunk 2]
```diff
@@ -491,34 +491,39 @@
     (let [{:keys [default-value]} argument-definition
           non-nullable? (non-null-kind? argument-definition)
           var-non-nullable? (non-null-kind? variable-def)
-          var-default-value (:default-value variable-def)]
+          var-default-value (:default-value variable-def)
+          variable-type (get-in argument-definition [:type :type])]
```
Comment (hlship) [inline src/com/walmartlabs/lacinia/parser.clj]: This assumes that the argument definition's type is `{:kind :root :type scalar-name}` and doesn't take into account simple things like a `:non-null` kind wrapper or (more involved) a `:list` kind wrapper.

File: test/com/walmartlabs/lacinia/custom_scalars_test.clj [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                                         :line 1}]
+                            :message "Invalid value for a scalar type."
+                            :query-path [:events
+                                         :lookup]
+                            :type :Event
+                            :value "1"}]}
+                 (execute schema q2 nil nil))
+              "should return error message"))))))
+
+(deftest custom-scalars-with-variables
```
Comment (hlship) [inline test/com/walmartlabs/lacinia/custom_scalars_test.clj]: An additional test or two, for `(not-null scalar-type)` and `(list scalar-type)` are needed.  I believe the code as currently implemented will not handle these cases.

File: test/com/walmartlabs/lacinia_test.clj [hunk 4]
```diff
@@ -686,87 +681,3 @@
                  :default-value "0001"}}}}}]
     (is-thrown [e (schema/compile schema-non-nullable-with-defaults)]
       (is (= (.getMessage e) "Field `id' of type `person' is both non-nullable and has a default value.")))))
-
-(deftest custom-scalars
```
Comment (hlship) [inline test/com/walmartlabs/lacinia_test.clj]: I'm very glad to see lacinia-test start to break up into smaller bits!

File: src/com/walmartlabs/lacinia/parser.clj (hunk 2 above)
Comment (bcarrell) [inline src/com/walmartlabs/lacinia/parser.clj]: Good call, still grokking the new type changes.

Comment (annapawlicka) [inline src/com/walmartlabs/lacinia/parser.clj]: I didn't understand new latest changes to types, but I do now. Fun times:
```
{:type {:kind :non-null, :type {:kind :list, :type {:kind :non-null, :type {:kind :root, :type :Date}}}}, :arg-name :between}
```

File: test/com/walmartlabs/lacinia/custom_scalars_test.clj (hunk 3 above)
Comment (annapawlicka) [inline test/com/walmartlabs/lacinia/custom_scalars_test.clj]: I added a bunch of tests with a combination of not-nulls and lists.

**Comment by annapawlicka** [comment]
@hlship @bcarrell I think I covered all cases, but let me know if there's something that I missed. The solution isn't as nice as I'd like it to be, but it's easiest for me to wrap my head around in this complex world of nested types and kinds.. I may give it another go once this new knowledge settles in my head.

**Comment by bcarrell** [comment]
Looks like this should work, to me.  I'll let @hlship merge if he's happy.

File: src/com/walmartlabs/lacinia/parser.clj [hunk 5]
```diff
@@ -465,12 +465,43 @@
...
-
 (defn ^:private non-null-kind?
   "Peeks at the kind of the provided def (field, argument, or variable) to see if it is :non-null"
   [any-def]
   (-> any-def :type :kind (= :non-null)))
 
+(defn ^:private process-result
+  "Checks result against variable kind, iterates over nested types, and applies respective
+  actions, if necessary, e.g. parse for custom scalars."
+  [schema result argument-definition arg-value]
```
Comment (hlship) [inline src/com/walmartlabs/lacinia/parser.clj]: An argument definition contains the argument type; I think (I'll check) that you are passing an argument-type here (the :type key of an argument definition).

Comment (hlship) [inline src/com/walmartlabs/lacinia/parser.clj]: Yes, on line 529.  So I might call this `argument-type`, and rename `next-type` to `nested-type` or `wrapped-type`.

File: src/com/walmartlabs/lacinia/parser.clj [hunk 6]
```diff
@@ -465,12 +465,43 @@
...
+
+      list-type?
+      (if (not (sequential? result))
+        (throw-exception (format "Variable %s doesn't contain the correct number of (nested) lists."
+                                 (q arg-value))
+                         {:variable-name arg-value})
+        (mapv #(process-result schema % next-type arg-value) result))
+
+      (and (some? result) (map? next-type))
+      (process-result schema result next-type arg-value)
```
Comment (hlship) [inline src/com/walmartlabs/lacinia/parser.clj]: How about `recur`?

**Comment by annapawlicka** [comment]
@hlship Done refactoring, back to you.

**Pull request merged**